### ISSUE TITLE
GitHub migration exceeds API rate limit

### ISSUE THREAD

**Issue by davidak** [go-gitea/gitea#13243]
- Gitea version (or commit ref): 1.13.0+rc1
- Git version: 2.20.1
- Operating system: debian stable 10.6
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [X] i get migration failed too, but can't look into your logs ;)

## Description

That should not happen. It should rather wait and inform the user, so the migration does not fail.

Related to https://github.com/go-gitea/gitea/issues/13241 and https://github.com/go-gitea/gitea/issues/13242

**Comment by lunny** [comment]
The problem will occur when you finished a migration and start another. On that situation, we haven't know if it's exceeds API rate limit.

**Comment by zeripath** [comment]
Maybe we should cache the github limit per user? We can even set the TTL to simply be the timeout.

**Comment by lunny** [comment]
We can try. And also when start two or more migrations per Github User, I think it also maybe failed.

**Comment by lafriks** [comment]
Doesn't github returns current rate limit in headers?

**Comment by lunny** [comment]
> Doesn't github returns current rate limit in headers?

If only one thread to do that, the returned current rate limit is meaningful. But when two or more threads requests github at the same time, it will not be useful.

**Comment by 6543** [comment]
look's like switching to GraphQL API will give us more requests per limit:
https://github.com/MichaelMure/git-bug/issues/315#issuecomment-786966651

**Comment by sbourdeauducq** [comment]
Is there any solution that f**** works in order to migrate Github repos right now, even a slow one?

Can I simply limit the numbers of threads to one so I don't get blocked by GitHub when migrating a repos with tens of thousands of issue comments?

Every single thing I've tried:
 * built-in gitea migration tool
 *  forgejo migration tool
 *  https://github.com/appleboy/github2gitea
 *  https://github.com/RayLabsHQ/gitea-mirror
 *  https://gitea.com/gitea/migrator

...is completely broken, and this is incredibly frustrating.

**Comment by lunny** [comment]
Have you tried internal migration feature?

**Comment by sbourdeauducq** [comment]
Yes. It's the first thing I tried. It gets rate limited and fails, even with multiple access tokens from several accounts.

**Comment by sbourdeauducq** [comment]
FWIW https://github.com/josegonzalez/python-github-backup is able to (slowly) download arbitrarily large amounts of data from GitHub, with a single access token. I've been using it for years to keep backups of important repos including issues, etc. Maybe that's a good starting point?

**Comment by sbourdeauducq** [comment]
In case this is useful to anyone, this works: https://git.m-labs.hk/fsagbuya/ghbackup-to-gitea

**Comment by lunny** [comment]
> In case this is useful to anyone, this works: [git.m-labs.hk/fsagbuya/ghbackup-to-gitea](https://git.m-labs.hk/fsagbuya/ghbackup-to-gitea)

Is the patch necessary? It seems Gitea have such columns of original author and original author id.

**Comment by fsagbuya** [comment]
> > In case this is useful to anyone, this works: [git.m-labs.hk/fsagbuya/ghbackup-to-gitea](https://git.m-labs.hk/fsagbuya/ghbackup-to-gitea)
> 
> Is the patch necessary? It seems Gitea have such columns of original author and original author id.

Yes, the patch is necessary. Gitea has the columns but the API only exposes them for reading. The native migration preserves original authors by writing directly to the database, but this tool relies solely on the [API](https://github.com/go-gitea/gitea/blob/3f46de82657a7ab04f043610639e13752687aad8/modules/structs/issue.go#L88) (for importing issues from [python-github-backup](https://github.com/josegonzalez/python-github-backup)). Without the patch, all imported issues would show the API caller as the author.

**Comment by rdehuyss** [comment]
I just want to mention that we also have trouble with this migrating a simple (larger) private repo from Github to gitea. We're moving because of copilot indexing.

**Comment by davidlovas** [comment]
I have implemented a GraphQL downloader for the GitHub migration after running into this same rate limit, as part of the metadata-mirror work on #18369 / #20311. Over REST every comment, review, reaction and timeline event is a separate call, so a large repo is hundreds of thousands of requests. GraphQL fetches an issue with its comments, reactions and timeline in one request, and a retry transport resumes across rate-limit windows so a large migration finishes unattended.

Tested by [mirroring the gitea repo](https://gitea.rexovas.com/rexovas/gitea) itself, as noted on #18369.

It's useful for any GitHub migration on its own, so I've opened it as a standalone PR: #38694 (`Closes #13243`).
